baseapp/grpcserver.go

```diff
@@ -2,67 +2,79 @@ package baseapp
-	// a new sdk.Context, and pass it into the query handler.
-	interceptor := func(grpcCtx context.Context, req interface{}, _ *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
-		// If there's some metadata in the context, retrieve it.
-		md, ok := metadata.FromIncomingContext(grpcCtx)
-		if !ok {
-			return nil, status.Error(codes.Internal, "unable to retrieve metadata")
-		}
+func (app *BaseApp) RegisterGRPCServer(clientCtx client.Context, server gogogrpc.Server) {
+	// Define an interceptor for all gRPC queries: this interceptor will route
+	// the query through the `clientCtx`, which itself queries Tendermint.
+	interceptor := func(grpcCtx context.Context, req interface{}, info *grpc.UnaryServerInfo, _ grpc.UnaryHandler) (interface{}, error) {
+		// Two things can happen here:
+		// 1. either we're broadcasting a Tx, in which case we call Tendermint's broadcast endpoint directly,
+		// 2. or we are querying for state, in which case we call ABCI's Query.
 
-		// Get height header from the request context, if present.
-		var height int64
-		if heightHeaders := md.Get(grpctypes.GRPCBlockHeightHeader); len(heightHeaders) > 0 {
-			height, err = strconv.ParseInt(heightHeaders[0], 10, 64)
-			if err != nil {
-				return nil, sdkerrors.Wrapf(
-					sdkerrors.ErrInvalidRequest,
-					"Baseapp.RegisterGRPCServer: invalid height header %q: %v", grpctypes.GRPCBlockHeightHeader, err)
-			}
-			if err := checkNegativeHeight(height); err != nil {
-				return nil, err
+		// Case 1. Broadcasting a Tx.
+		if client.IsGRPCBroadcastTx(info.FullMethod) {
+			reqProto, ok := req.(*tx.BroadcastTxRequest)
+			if !ok {
+				return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "expected %T, got %T", (*tx.BroadcastTxRequest)(nil), req)
 			}
+
+			return client.TxServiceBroadcast(grpcCtx, clientCtx, reqProto)
 		}
 
-		// Create the sdk.Context. Passing false as 2nd arg, as we can't
-		// actually support proofs with gRPC right now.
-		sdkCtx, err := app.createQueryContext(height, false)
+		// Case 2. Querying state.
+		inMd, _ := metadata.FromIncomingContext(grpcCtx)
+		abciRes, outMd, err := client.RunGRPCQuery(clientCtx, grpcCtx, info.FullMethod, req, inMd)
 		if err != nil {
 			return nil, err
 		}
 
-		// Attach the sdk.Context into the gRPC's context.Context.
-		grpcCtx = context.WithValue(grpcCtx, sdk.SdkContextKey, sdkCtx)
+		// We need to know the return type of the grpc method for
+		// unmarshalling abciRes.Value.
+		//
+		// When we call each method handler for the first time, we save its
+		// return type in the `returnTypes` map (see the method handler in
+		// `grpcrouter.go`). By this time, the method handler has already run
+		// at least once (in the RunGRPCQuery call), so we're sure the
+		// returnType maps is populated for this method. We're retrieving it
+		// for decoding.
+		returnType, found := app.GRPCQueryRouter().returnTypes[info.FullMethod]
```

Comment:
Instead of using reflections, I would propose something else.

All return types are `proto.Message`

Instead of having a map:

```
methodName -> reflect.Type
``` 

I'm not sure where is the best place to create that functions - probably each module will need to do it itself when registering interfaces? So a bit more work, and more verbose. Not for this task, but idea for the future. If you think this make sense then we can create an issue for this.
we can use factory functions:
```
methodName -> func() proto.Message
```
Example:
```

Reply:
we could do that already, but still using reflection:

```diff
- qrt.returnTypes[fqName] = reflect.TypeOf(res)
+ qrt.returnTypes[fqName] = func() {
+   typ := reflect.TypeOf(res)
+   return reflect.New(typ.Elem()).Interface().(proto.Message)
+ }
```

Both are equivalent in my eyes, I don't have a preference. happy to switch to factory functions (with reflection inside) if you find them clearer.

Reply:
I was thinking more about registering this methods in the modules when registering the query server, rather than `baseapp`. So that would be before even running the query server. It's more declarative - so a bit more code, but we won't use reflection.

Reply:
shall we create a task to discuss it further?

Reply:
I created one: https://github.com/cosmos/cosmos-sdk/issues/8588